neuraxle/runners.py

```diff
@@ -0,0 +1,277 @@
+                                                                 data_inputs,
+                                                                 expected_outputs)
+        else:
+            return step.fit_transform(data_inputs, expected_outputs)
+
+    def fit_transform_steps_left_to_do_recursive(self, step, data_inputs, expected_outputs):
+        """
+        After loading the last checkpoint, recursively fit transform a step that contains steps
+        :param data_inputs: the data input to fit on
+        :param expected_outputs: the expected data output to fit on
+        :param step: pipeline step
+        :return: the pipeline itself
+        """
+        checkpoint, index = self.find_starting_step_and_data_inputs(
+            steps_as_tuple=step.steps_as_tuple,
+            data_inputs=data_inputs,
+            expected_outputs=expected_outputs
+        )
+
+        checkpoint_data_input, checkpoint_expected_outputs = checkpoint
+        if checkpoint_data_input is not None \
+                and checkpoint_expected_outputs is not None:
+            steps_left_to_do_recursive = step.steps_as_tuple[index:]
+            return self.fit_transform_steps_left_to_do(checkpoint_data_input,
+                                                       checkpoint_expected_outputs,
+                                                       steps_left_to_do_recursive)
+
+        return step.fit_transform(data_inputs, expected_outputs)
+
+    def transform(self, data_inputs):
+        """
+        After loading the last checkpoint, transform each pipeline steps
+        :param data_inputs: the data input to fit on
+        :return: transformed data inputs
+        """
+        new_steps_as_tuple = self.create_checkpoint_names(self.steps_as_tuple)
+        self.steps_as_tuple = new_steps_as_tuple
+
+        (data_inputs, _), index_starting_step = self.find_starting_step_and_data_inputs(
+            steps_as_tuple=self.steps_as_tuple,
+            data_inputs=data_inputs
+        )
+
+        steps_left_to_do = self.steps_as_tuple[index_starting_step:]
+        data_inputs = self.transform_steps_left_to_do(data_inputs, steps_left_to_do)
+
+        return data_inputs
+
+    def transform_steps_left_to_do(self, data_inputs, steps_left_to_do):
+        """
+        After loading the last checkpoint, transform each pipeline steps left to do
+        :param data_inputs: the data input to transform on
+        :param steps_left_to_do: steps left to do
+        :return: the pipeline itself
+        """
+        for step_name, step in steps_left_to_do:
+            if isinstance(step, TruncableSteps) or isinstance(step, MetaStepsMixin):
+                data_inputs = self.transform_steps_left_to_do_recursive(step, data_inputs)
+            else:
+                data_inputs = step.transform(data_inputs)
+
+        return data_inputs
+
+    def transform_steps_left_to_do_recursive(self, step, data_inputs):
```

Comment:
What happen if we have different pipeline runners in the steps below (whithin / nested) in the current step, do we want to use the current pipeline runner, or let the sub-step use its own runner? I think we need to let the step use its own runner (Tell Don't Ask principle, again). This TDA principle is crucial. 

For instance, if we have a new runner in a pipeline that is a step of the current pipeline, this step should use its own runner and the current runner should not highjack the step that it contains.

Reply:
I wanted to do that at first, but it became very complicated. I am not sure what to do here, but the parent runner has priority right now

Reply:
Could you please check the code that you deleted where I had some kind of pseudocode to do that? (You might have missed some of my comments about that because GitHub collapsed some of the comments above, you need to click something to unwrap some of them).

Reply:
It's the dirty code with lots of TODOs that was in the pipeline.py file.